File: js/maths/primenumber.js
// problem: given a natural number 'n', determine if the number is a prime number or not;

// prime number is a natural numbers greater than 1 and that can only be divided by 1 and itself
// O(n)
function prime( n ) {
    
    if ( n < 2 ) return false

    for ( let i = 2; i < n; i++ ) {
        if (n % i === 0) {
            return false;
        }
    }
    return true
}

// console.log( prime( 11 ) );



// optimized prime
// O(sqrt(n))
function prime2( n ) {
     
    if ( n < 2 ) return false

    for ( let i = 2; i <= Math.sqrt(n); i++ ) {
        if (n % i === 0) {
            return false;
        }
    }
    return true
}

console.log( prime( 10 ) );
console.log( prime( 11 ) );
console.log( prime( 9 ) );
console.log( prime( 8 ) );
console.log( prime( 4 ) );

File: java/maths/PrimeNumber.java
public class PrimeNumber {
    

    static boolean primeNum(int n) {
        if (n < 2)
            return false;

        for (int i = 2; i < n; i++) {
            if (n % i == 0) {
                return false;
            }
        }
        return true;
    }
    
    // O(logn)
    static boolean primeNum2(int n) {
        if (n < 2)
            return false;

        for (int i = 2; i <= Math.sqrt(n); i++) {
            if (n % i == 0) {
                return false;
            } 
        }
        return true;
    }
    
    public static void main(String[] args) {
        System.out.println(primeNum2(11));
        System.out.println(primeNum2(9));
        System.out.println(primeNum2(8));
        System.out.println(primeNum2(4));
    }
}
